tools/export/cmake/__init__.py

```diff
@@ -0,0 +1,225 @@
+"""
+mbed SDK
+Copyright (c) 2011-2016 ARM Limited
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import re
+import shutil
+from os import remove
+from os.path import splitext, basename, exists
+from subprocess import Popen, PIPE
+
+from jinja2.exceptions import TemplateNotFound
+
+from tools.export.exporters import Exporter
+from tools.utils import NotSupportedException
+
+
+class CMake(Exporter):
+    """Generic CMake template that mimics the behavior of the python build
+    system
+    """
+
+    TEMPLATE = 'CMakeLists.txt'
+
+    MBED_CONFIG_HEADER_SUPPORTED = True
+
+    PREPROCESS_ASM = False
+
+    @classmethod
+    def is_target_supported(cls, target_name):
+        return True
+
+    def generate(self):
+        """Generate the CMakefiles.txt
+        """
+        self.resources.win_to_unix()
+
+        sources = set(self.resources.c_sources + \
+                      self.resources.cpp_sources + \
+                      self.resources.s_sources + \
+                      self.resources.headers)
+
+        libnames = [l[:-4] for l in self.resources.lib_refs]
+        libs = {re.sub(r'^[.]/', '', l): sorted([f for f in sources if f.startswith(l)]) for l in libnames}
+        libs = {k: v for k, v in libs.items() if len(v) != 0}
+        srcs = sorted([f for f in sources if f not in [item for sublist in libs.values() for item in sublist]])
```

Comment:
This looks like it might be a check for uniqueness. `set()`, used above, already guarantees uniqueness, so I don't think this is needed. Further, it's very hard to follow exactly what this is doing.

Reply:
I will take a look and update the PR. Some of your comments obviously relay to my lazyness...

Reply:
explanation in the code. python has not subtract list from list, so its done like this.